src/tfs/tfs.c

```diff
@@ -783,6 +937,114 @@ void link(tuple dir, fsfile f, buffer name)
 }
 #endif
 
+void filesystem_alloc(filesystem fs, tuple t, long offset, long len,
+        boolean keep_size, fs_status_handler completion)
+{
+    fsfile f = table_find(fs->files, t);
+    assert(f);
+    tuple extents = table_find(t, sym(extents));
+    if (!extents) {
+        apply(completion, f, FS_STATUS_NOENT);
+        return;
+    }
+    range q = irange(offset, offset + len);
+    tfs_debug("%s: t %v, q %R%s\n", __func__, t, q,
+            keep_size ? " (keep size)" : "");
+
+    struct rangemap new_rm;
+    list_init(&new_rm.root);
+    fs_status status = FS_STATUS_OK;
+
+    u64 lastedge = q.start;
+    list_foreach(&f->extentmap->root, l) {
+        rmnode curr = struct_from_list(l, rmnode, l);
+        u64 edge = curr->r.start;
+        range i = range_intersection(irange(lastedge, edge), q);
+        if (range_span(i)) {
+            if (!add_extents(fs, i, &new_rm)) {
+                status = FS_STATUS_NOSPACE;
+                goto error;
+            }
+        }
+        lastedge = curr->r.end;
+    }
+
+    /* check for a gap between the last node and q.end */
+    range i = range_intersection(irange(lastedge, q.end), q);
+    if (range_span(i)) {
+        if (!add_extents(fs, i, &new_rm)) {
+            status = FS_STATUS_NOSPACE;
+            goto error;
+        }
+    }
+
+    boolean update_size;
+    if (!keep_size && (q.end > fsfile_get_length(f))) {
+        update_size = true;
+    } else {
+        update_size = false;
+    }
+    if (!update_size && list_empty(&new_rm.root)) {
+        apply(completion, f, FS_STATUS_OK);
+        return;
+    }
+    merge m = allocate_merge(fs->h,
```

Comment:
See other comment about taking a merge reference for the issuing operation itself.